tensorflow_quantum/datasets/spin_system.py

```diff
@@ -0,0 +1,332 @@
+# Copyright 2020 The TensorFlow Quantum Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Quantum datasets for quantum many-body spin systems."""
+
+from collections import namedtuple
+import os
+
+import numpy as np
+import sympy
+import cirq
+import tensorflow as tf
+
+SpinSystemInfo = namedtuple(
+    "SpinSystemInfo",
+    [
+        "g",  # Numpy `float` order parameter.
+        "gs",  # Numpy `complex` array ground state wave function from exact
+        # diagonalization.
+        "gs_energy",  # Numpy `float` ground state energy from exact
+        # diagonalization.
+        "res_energy",  # Python `float` residual between the  circuit energy and
+        # the exact energy from exact diagonalization.
+        "fidelity",  # Python `float` overlap between the circuit state and the
+        # exact ground state from exact diagonalization.
+        "params",  # Dict with Python `str` keys and Numpy`float` values.
+        # Contains $M \times P parameters. Here $M$ is the number of
+        # parameters per circuit layer and $P$ the circuit depth.
+        "circuit"  # Variational `cirq.Circuit` quantum circuit with Sympy
+        # parameters
+    ])
+
+
+def unique_name():
+    """Generator to generate an infinite number of unique names.
+
+    Yields:
+        Python `str` of the form "theta_<integer>"
+
+    """
+    num = 0
+    while True:
+        yield "theta_" + str(num)
+        num += 1
+
+
+def download_spin_data(nspins, system_name, data_dir):
```

Comment:
Might want to change to `_download_spin_data` since we don't intend for this to be used outside of this module.